libsoftwarecontainer/src/gateway/devicenodegateway.cpp

```diff
@@ -50,22 +52,25 @@ bool DeviceNodeGateway::activateGateway()
 {
     for (auto &dev : m_devList) {
         log_info() << "Mapping device " << dev.name;
+        const int mode = std::atoi(dev.mode.c_str());
 
         if (dev.major.length() != 0) {
-            auto success = executeInContainer("mknod " + dev.name + " c " + dev.major + " " + dev.minor);
-            if (!isError(success)) {
-                success = executeInContainer("chmod " + dev.mode + " " + dev.name);
-            } else {
+            const int majorVersion = std::atoi(dev.major.c_str());
+            const int minorVersion = std::atoi(dev.minor.c_str());
+
+            // mknod dev.name c dev.major dev.minor
+            if (mknod(dev.name.c_str(), S_IFCHR | mode, makedev(majorVersion, minorVersion)) != 0) {
                 log_error() << "Failed to create device " << dev.name;
                 return false;
             }
         } else {
             // No major & minor numbers specified => simply map the device from the host into the container
             getContainer()->mountDevice(dev.name);
 
-            // TODO : check if it is fine to authorize write access systematically
-            std::string cmd = StringBuilder() << "chmod o+rwx " << dev.name;
-            getContainer()->executeInContainer(cmd);
+            if (chmod(dev.name.c_str(), (S_IRUSR + S_IWUSR + S_IXUSR) | mode) != 0) {
+                log_error() << "Could not 'chmod o+rwx' mounted dev" << dev.name;
```

Comment:
This is not equivalent to o+rwx, this is equivalent to u+rwx. Which should it be?